lib/devicetrust/native/device_darwin.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package native
+
+// #cgo CFLAGS: -Wall -xobjective-c -fblocks -fobjc-arc -mmacosx-version-min=10.13
+// #cgo LDFLAGS: -framework CoreFoundation -framework Foundation -framework IOKit -framework Security
+// #include <stdint.h>
+// #include <stdlib.h>
+// #include "device_darwin.h"
+import "C"
+
+import (
+	"crypto/sha256"
+	"crypto/x509"
+	"fmt"
+	"unsafe"
+
+	"github.com/google/uuid"
+	"github.com/gravitational/trace"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/lib/apple"
+)
+
+func enrollDeviceInit() (*devicepb.EnrollDeviceInit, error) {
+	cred, err := deviceKeyCreate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cd, err := collectDeviceData()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &devicepb.EnrollDeviceInit{
+		CredentialId: cred.Id,
+		DeviceData:   cd,
+		Macos: &devicepb.MacOSEnrollPayload{
+			PublicKeyDer: cred.PublicKeyDer,
+		},
+	}, nil
+}
+
+func deviceKeyCreate() (*devicepb.DeviceCredential, error) {
+	newID := uuid.NewString()
+	newIDC := C.CString(newID)
+	defer C.free(unsafe.Pointer(newIDC))
+
+	var pubKeyC C.PublicKey
+	defer func() {
+		C.free(unsafe.Pointer(pubKeyC.id))
+		C.free(unsafe.Pointer(pubKeyC.pub_key))
+	}()
+
+	if res := C.DeviceKeyCreate(newIDC, &pubKeyC); res != 0 {
+		return nil, trace.Wrap(fmt.Errorf("creating device key: status %d", res))
+	}
+
+	return pubKeyToCredential(pubKeyC)
```

Comment:
The fact that we pass `pubKeyC` to other function combined with the free's make the allocation analysis non-trivial. My suggestions:
- introduce new type `type publicKey struct { id  string; raw []byte; }`
- unmarshal `pubKeyC` to `publicKey` directly in `deviceKeyCreate`
- change the method signatures to `deviceKeyCreate() (*publicKey, error)` and `pubKeyToCredential(pubKey publicKey) (*devicepb.DeviceCredential, error)`
- call `pubKeyToCredential` from `enrollDeviceInit` instead of tail call from `deviceKeyCreate`

This makes it easier to see that all unmanaged memory is truly freed as soon as `deviceKeyCreate` returns. Currently you need to analyze `pubKeyToCredential` to see if no funky business takes place.